### PR TITLE
Mark API strings for translation

### PR DESCRIPTION
**Pull request by jerabekjiri**
Issue: [AAH-830](https://issues.redhat.com/browse/AAH-830)
Marked most of the strings for translation and added gettext_lazy import, but I'm not sure about these translations:
 - `self.stdout.write`
 - `self.status_messages.append`
 - `parser.add_argument('name', type=slugify, help=_("Remote Name"))`
 - `log.info, log.debug, log.warning`

### CODE REVIEW

File: galaxy_ng/app/api/v3/serializers/collection.py [hunk 1]
```diff
@@ -142,7 +136,7 @@ def to_internal_value(self, data):
             filename_tuple = parse_collection_filename(filename)
         except ValueError as exc:
             errors['filename'] = _get_error_details(exc, default_code='invalid')
-            log.error('CollectionUploadSerializer validation of filename failed: %s',
+            log.error(_('CollectionUploadSerializer validation of filename failed: %s'),
```
Comment (rochacbruno) [inline galaxy_ng/app/api/v3/serializers/collection.py:139]: Log entries doesn't need translation as it is not exposed to the user, logs are more useful for devs and ops so it is better to search for occurrences in English.

**Review by rochacbruno — CHANGES_REQUESTED** [review]
- Do not translate log messages
- Your commits missed `Issue: AAH-999` (with issue number)
- All commits must be signed

File: galaxy_ng/app/common/parsers.py [hunk 2]
```diff
@@ -19,7 +20,7 @@ def parse(self, stream, media_type=None, parser_context=None):
         new_stream.write(body_parts[1])
 
         if not body_parts[2].startswith(b'\r\n'):
-            log.warning('Malformed multipart body user-agent: %s',
+            log.warning(_('Malformed multipart body user-agent: %s'),
```
Comment (awcrosby) [inline galaxy_ng/app/common/parsers.py:23]: ditto about log msg per bruno's comment

File: galaxy_ng/app/management/commands/assign-permission.py [hunk 3]
```diff
@@ -4,10 +4,10 @@
...
 from django.core.management import BaseCommand, CommandError
 from django.core.exceptions import ObjectDoesNotExist
+from django.utils.translation import gettext_lazy as _
 
 User = get_user_model()
 
-
 class Command(BaseCommand):
     """
     Django management command for assigning permissions to groups.
```
Comment (awcrosby) [inline galaxy_ng/app/management/commands/assign-permission.py:13]: if the goal of translating the api is ensure translated strings get passed to the UI, then we probably don't need to translate the management commands `galaxy_ng/app/management/commands/*` since they are not part of the api

**Review by awcrosby — COMMENTED** [review]
Looking good so far. Added a couple comments. Also we'll need to add back in the blank lines.

**Comment by jerabekjiri** [comment]
Fixed requested changes and errors :)
Should we be translating `@extend_schema` as it's used for swagger-ui?
```
    @extend_schema(summary=_("Retrieve collection version"),
                   responses={200: serializers.CollectionVersionDetailSerializer})
```
And what about `help_text`, not sure here?
```
    proxy_password = serializers.CharField(
        help_text=_("Password for proxy authentication."),
        allow_null=True,
        required=False,
        ...
```
check pls @rochacbruno :)

**Comment by rochacbruno** [comment]
@jerabekjiri yes, extend schema and help text are user-facing texts so must be translated.

**Comment by jerabekjiri** [comment]
Squashed commits and I think this PR should be ready.

The question is how to switch languages? 
I think for this case, we can use `Accept-Language HTTP header` and set it like that  `Accept-Language: fr` for example.

Also, @chouseknecht had an idea to ensure that newly added strings will be marked for translation.
I'm not sure this is perfectly possible, but I think we could make a set of words (mentioned above: `help_text, @extend_schema, ValidationError, NotImplementedError`, etc...) which their strings always must be translated and then run it in `./ci/scripts/check_gettext.sh`, but I think that's for entirely new PR.

can you check pls @rochacbruno, @awcrosby :)

File: galaxy_ng/app/api/base.py [hunk 4]
```diff
@@ -23,7 +23,7 @@
 
 class _MustImplementPermission(permissions.BasePermission):
     def has_permission(self, request, view):
-        raise NotImplementedError("subclass must implement permission_classes")
+        raise NotImplementedError(_("subclass must implement permission_classes"))
```
Comment (rochacbruno) [inline galaxy_ng/app/api/base.py:26]: `NotImplementedError` is raised on the backend but never go to the user level.

Those messages should not be translated.

It is different on ValidationError because DRF raises ValidationError to the user.

**Review by rochacbruno — CHANGES_REQUESTED** [review]

**Comment by rochacbruno** [comment]
@jerabekjiri we can use a regex like this https://regex101.com/r/Kem4Cj/2 to check for untranslated strings, this is tricky because some are impossible to catch.

**Comment by jerabekjiri** [comment]
> we can use a regex like this https://regex101.com/r/Kem4Cj/2 to check for untranslated strings, this is tricky because some are impossible to catch.

that's what I thought, we can't really track variables like [errmsg](https://github.com/ansible/galaxy_ng/blob/3c4693e6fe13d116e4cd00fff8d86acec7970ec2/galaxy_ng/app/api/ui/serializers/synclist.py#L41) that are passed to ValidationError for example (or would be a lot more complicated).

https://regex101.com/r/P2RAfB/1/ added more keywords:
- ValueError
- Http404
- ConflictError
- APIException
- NotFound
- RuntimeError
- and multiple row command &#8595;
 
```
ValidationError(
         detail={
                'groups': 'object_permissions field is required'})
```

File: galaxy_ng/app/api/v3/viewsets/collection.py [hunk 5]
```diff
@@ -274,8 +275,8 @@ def get(self, request, *args, **kwargs):
             )
 
         metrics.collection_artifact_download_failures.labels(status=response.status_code).inc()
-        raise APIException('Unexpected response from content app. '
-                           f'Code: {response.status_code}.')
+        raise APIException(_('Unexpected response from content app. '
+                           f'Code: {response.status_code}.'))
```
Comment (awcrosby) [inline galaxy_ng/app/api/v3/viewsets/collection.py:279]: this has an f-string which should be removed, looks like ci didn't catch it because it is on multiple lines

File: galaxy_ng/app/management/commands/create-remote.py [hunk 6]
```diff
@@ -78,7 +79,8 @@ def add_arguments(self, parser):
             "-r",
             "--requirements_file",
             type=self.load_requirements_file,
-            help="Either an abs path ending in .yaml|.yml to be loaded or the literal YAML string."
+            help="Either an abs path ending in .yaml|.yml "
```
Comment (awcrosby) [inline galaxy_ng/app/management/commands/create-remote.py:82]: minor: probably can undo changes to management command files since this they are not related to the pr

File: galaxy_ng/app/api/v3/viewsets/collection.py [hunk 7]
```diff
@@ -170,7 +169,9 @@ def _check_path_matches_expected_repo(path, filename_ns):
         if INBOUND_REPO_NAME_FORMAT.format(namespace_name=filename_ns) == repo_name:
             return
         raise NotFound(
-            f'Path does not match: "{INBOUND_REPO_NAME_FORMAT.format(namespace_name=filename_ns)}"')
+            _('Path does not match: '
+              f'"{INBOUND_REPO_NAME_FORMAT.format(namespace_name=filename_ns)}"')
```
Comment (awcrosby) [inline galaxy_ng/app/api/v3/viewsets/collection.py:173]: this code has a mix of f-strings and .format(), and it looks like f-strings are not supported with gettext

**Review by awcrosby — CHANGES_REQUESTED** [review]

File: CHANGES/830.misc [hunk 8]
```diff
@@ -0,0 +1 @@
+Mark API strings for translation and add 'from django.utils.translation import gettext_lazy as _'
```
Comment (awcrosby) [inline CHANGES/830.misc:1]: Suggest renaming pr title to `Mark API strings for translation` to match part of this changelog msg. We typically reuse pr title for commit title when squashing and merging

File: galaxy_ng/app/api/v3/viewsets/collection.py [hunk 9]
```diff
@@ -274,8 +275,8 @@ def get(self, request, *args, **kwargs):
             )
 
         metrics.collection_artifact_download_failures.labels(status=response.status_code).inc()
-        raise APIException('Unexpected response from content app. '
-                           f'Code: {response.status_code}.')
+        raise APIException(_('Unexpected response from content app. '
+                             'Code: %s.' % response.status_code))
```
Comment (awcrosby) [inline galaxy_ng/app/api/v3/viewsets/collection.py:279]: It seems most places the string variable is outside the `gettext_lazy` method, but at least 2 places in this file have it inside... should it always be outside?
inside: `gettext_lazy('Unexpected response from content app. Code: %s.' % response.status_code)`
outside: `gettext_lazy('Collection %s not found') % version_str`

@jerabekjiri do you regenerate the `locale/*/LC_MESSAGES/django.po` files after each code edit? That might be a good test to see if it is supported. For example the .po item related to this is: `msgid "Unexpected response from content app. "`

Line 278 & 279 could probably fit on one line and be easier to read / edit in future.

Comment (jerabekjiri) [reply galaxy_ng/app/api/v3/viewsets/collection.py:279]: > It seems most places the string variable is outside the gettext_lazy method, but at least 2 places in this file have it inside... should it always be outside?

Seems both versions work and generate the right strings, but I'll change it to outside, according to Django documentation.

I'll regenerate the messages and push changed files, I forgot to renew them :)

Comment (awcrosby) [reply galaxy_ng/app/api/v3/viewsets/collection.py]: thanks, looks good

**Review by awcrosby — APPROVED** [review]

**Comment by awcrosby** [comment]
LGTM, I approved, but we need to get the `ci.ext.devshift.net PR build` CI check to run... I'm not sure if any of the `/retest` comments worked earlier. I feel PRs with multiple commits over time sometimes have the CI issues (wild guess)... maybe squashing into 1 commit could help.

**Review by rochacbruno — APPROVED** [review]
It would be nice to generate some translations and write a simple test case

**Comment by jerabekjiri** [comment]
Added localization test to check the translation, but seems like I will have to add the `compilemessages` command to generate .mo files for the translations.

**Comment by rochacbruno** [comment]
@jerabekjiri you can add a call to that script on the file `.github/workflows/scripts/post_before_script.sh` 

You can add to that file:

This runs before the tests are axecuted.

`.github/workflows/scripts/post_before_script.sh` (this file doesn't exist yet, but will be called once it exists)
```bash
#!/usr/bin/env bash

set -mveuo pipefail
source .github/workflows/scripts/utils.sh
cmd_prefix bash -c "django-admin compilemessages"
```

File: .github/workflows/scripts/before_script.sh [hunk 10]
```diff
@@ -10,9 +10,9 @@
 # make sure this script runs at the repo root
 cd "$(dirname "$(realpath -e "$0")")"/../../..
 
-set -euv
-
+set -mveuo pipefail
 source .github/workflows/scripts/utils.sh
+cmd_prefix bash -c "django-admin compilemessages"
```
Comment (rochacbruno) [inline .github/workflows/scripts/before_script.sh:15]: You can't edit this file directly, you need to create a new file called `.github/workflows/scripts/post_before_script.sh` and place the script you need there.

This file is managed by pulp plugin template and will be overriden.

**Review by rochacbruno — CHANGES_REQUESTED** [review]

File: .github/workflows/scripts/before_script.sh (hunk 10 above)
Comment (jerabekjiri) [reply .github/workflows/scripts/before_script.sh]: sorry, I misread the last comment, fixed it

**Comment by jerabekjiri** [comment]
From logs, .po files are being processed now, but the translation from files still doesn't work, I had a problem with it on localhost and solved it by adding, so that might be the cause
```
LOCALE_PATHS = (os.path.join('/src/galaxy_ng/locale'), )
```

File: galaxy_ng/app/settings.py [hunk 11]
```diff
@@ -115,6 +116,23 @@
...
+
+LANGUAGES = [
+    ('ja', 'Japanese'),
+    ('nl', 'Dutch'),
+    ('fr', 'French'),
+    ('es', 'Spanish'),
+    ('zh-hans', 'Chinese'),
+]
+
+LOCALE_PATHS = (os.path.join('/src/galaxy_ng/locale'), )
```
Comment (rochacbruno) [inline galaxy_ng/app/settings.py:134]: @jerabekjiri 

I think you need to move the `locale` folder from the root of repo to inside the `galaxy_ng` module folder.

`mv locale galaxy_ng/locale`

then this setting can be

```suggestion
from pathlib import Path  # move this line to the top of the file
BASE_DIR = Path(__file__).absolute().parent
LOCALE_PATHS = (os.path.join(BASE_DIR, 'locale'),)
```

File: MANIFEST.in [hunk 12]
```diff
@@ -6,3 +6,4 @@ include pyproject.toml
 include requirements/requirements.common.txt
 graft galaxy_ng/app/webserver_snippets
 graft galaxy_ng/app/static
+recursive-include locale *.po
```
Comment (rochacbruno) [inline MANIFEST.in:9]: After you move the `locale` to `galaxy_ng/locale`

```suggestion
graft galaxy_ng/locale
```

`graft` takes the whole folder and adds to the distribution binaries.

**Review by rochacbruno — CHANGES_REQUESTED** [review]

File: galaxy_ng/app/settings.py (hunk 11 above)
Comment (rochacbruno) [reply galaxy_ng/app/settings.py:134]: Here an example https://github.com/ansible/awx/tree/devel/awx/locale and https://github.com/ansible/awx/blob/9da383fe2d3bd803fd6e6766246e3b4f6566bd96/awx/settings/defaults.py#L140

**Review by rochacbruno — APPROVED** [review]

**Comment by jerabekjiri** [comment]
@rochacbruno you figured it out, tests are passing now, thanks a lot for the help :)

**Comment by awcrosby** [comment]
@jerabekjiri you will need to sign all your commits. after that I think we can merge.

**Comment by jerabekjiri** [comment]
Squashed and signed the commit
